util/linuxfw/iptables.go

```diff
@@ -0,0 +1,821 @@
+	startOff := p.offset
+
+	iptip, err := p.parseIPTIP()
+	if err != nil {
+		return entry{}, fmt.Errorf("parsing IPTIP: %w", err)
+	}
+
+	ret := entry{
+		Offset: startOff,
+		IP:     iptip,
+	}
+
+	// Must have space for the rest of the members
+	if err := p.assertLen(28); err != nil {
+		return entry{}, err
+	}
+
+	ret.NFCache = native.Endian.Uint32(p.getBytes(4))
+	targetOffset := int(native.Endian.Uint16(p.getBytes(2)))
+	nextOffset := int(native.Endian.Uint16(p.getBytes(2)))
+	/* Comeback = */ p.getBytes(4)
+	ret.PacketCount = native.Endian.Uint64(p.getBytes(8))
+	ret.ByteCount = native.Endian.Uint64(p.getBytes(8))
+
+	// Must have at least enough space in our buffer to get to the target;
+	// doing this here means we can avoid bounds checks in parseMatches
+	if err := p.assertLen(targetOffset - p.offset); err != nil {
+		return entry{}, err
+	}
+
+	// Matches are stored between the end of the entry structure and the
+	// start of the 'targets' structure.
+	ret.Matches, err = p.parseMatches(targetOffset)
+	if err != nil {
+		return entry{}, err
+	}
+
+	if targetOffset > 0 {
+		if err := p.setOffset(targetOffset, nextOffset); err != nil {
+			return entry{}, err
+		}
+
+		ret.Target, err = p.parseTarget(nextOffset)
+		if err != nil {
+			return entry{}, fmt.Errorf("parsing target: %w", err)
+		}
+	}
+
+	if err := p.setOffset(nextOffset, -1); err != nil {
+		return entry{}, err
+	}
+
+	return ret, nil
+}
+
+type iptip struct {
+	Src                 netip.Addr
+	Dst                 netip.Addr
+	SrcMask             netip.Addr
+	DstMask             netip.Addr
+	InputInterface      string
+	OutputInterface     string
+	InputInterfaceMask  []byte
+	OutputInterfaceMask []byte
+	Protocol            uint16
+	Flags               uint8
+	InverseFlags        uint8
+}
+
+var protocolNames = map[uint16]string{
+	unix.IPPROTO_ESP:    "esp",
+	unix.IPPROTO_GRE:    "gre",
+	unix.IPPROTO_ICMP:   "icmp",
+	unix.IPPROTO_ICMPV6: "icmpv6",
+	unix.IPPROTO_IGMP:   "igmp",
+	unix.IPPROTO_IP:     "ip",
+	unix.IPPROTO_IPIP:   "ipip",
+	unix.IPPROTO_IPV6:   "ip6",
+	unix.IPPROTO_RAW:    "raw",
+	unix.IPPROTO_TCP:    "tcp",
+	unix.IPPROTO_UDP:    "udp",
+}
+
+func (ip iptip) String() string {
+	var sb strings.Builder
+	sb.WriteString(`{`)
+
+	formatAddrMask := func(addr, mask netip.Addr) string {
+		if pref, ok := netaddr.FromStdIPNet(&net.IPNet{
+			IP:   addr.AsSlice(),
+			Mask: mask.AsSlice(),
+		}); ok {
+			return fmt.Sprint(pref)
+		}
+		return fmt.Sprintf(`%s/%s`, addr, mask)
+	}
+
+	fmt.Fprintf(&sb, `Src:%s`, formatAddrMask(ip.Src, ip.SrcMask))
+	fmt.Fprintf(&sb, `, Dst:%s`, formatAddrMask(ip.Dst, ip.DstMask))
+
+	translateMask := func(mask []byte) string {
+		var ret []byte
+		for _, b := range mask {
+			if b != 0 {
+				ret = append(ret, 'X')
+			} else {
+				ret = append(ret, '.')
+			}
+		}
+		return string(ret)
+	}
+
+	if ip.InputInterface != "" {
+		fmt.Fprintf(&sb, `, InputInterface:%s/%s`, ip.InputInterface, translateMask(ip.InputInterfaceMask))
```

Comment:
not a fan of backticks for string literals except 1) in regexps always, even when not needed, 2) when really needed because of newlines or lots of double quotes

Reply:
Switched everything to regular double-quotes.